src/metadata/metadata.rs

```diff
@@ -0,0 +1,207 @@
+use crate::spec::Spec;
+
+pub type Date = NaiveDate;
+
+#[derive(Debug, Default, Clone)]
+pub struct MetadataManager {
+    pub has_metadata: bool,
+    pub abs: Option<Vec<String>>,
+    pub canonical_url: Option<String>,
+    pub date: Option<Date>,
+    pub ed: Option<String>,
+    pub editors: Option<Vec<String>>,
+    pub group: Option<String>,
+    pub level: Option<String>,
+    pub shortname: Option<String>,
+    pub raw_status: Option<String>,
+    pub title: Option<String>,
+}
+
+impl MetadataManager {
+    pub fn new() -> MetadataManager {
+        MetadataManager {
+            has_metadata: false,
+            ..Default::default()
+        }
+    }
+
+    pub fn join_all(sources: &[&MetadataManager]) -> MetadataManager {
+        let mut mm = MetadataManager::new();
+        for source in sources {
+            mm.join((*source).clone());
+        }
+        mm
+    }
+
+    pub fn add_data(&mut self, key: &str, val: &String, line_num: Option<u32>) {
+        let mut key = key.trim().to_string();
+
+        if key != "ED" && key != "TR" && key != "URL" {
+            key = titlecase(&key);
+        }
+
+        match key.as_str() {
+            "Abstract" => {
+                let val = parse::parse_vec(val);
+                self.abs.join(Some(val));
+            }
+            "Canonical Url" => {
+                let val = val.clone();
+                self.canonical_url.join(Some(val));
+            }
+            "Date" => {
+                let val = match parse::parse_date(val) {
+                    Ok(val) => val,
+                    Err(_) => {
+                        die!("The \"Date\" field must be in the format YYYY-MM-DD."; line_num)
+                    }
+                };
+                self.date.join(Some(val));
+            }
+            "ED" => {
+                let val = val.clone();
+                self.ed.join(Some(val));
+            }
+            "Editor" => {
+                let val = parse::parse_editor(val);
+                self.editors.join(Some(val));
+            }
+            "Group" => {
+                let val = val.clone();
+                self.group.join(Some(val));
+            }
+            "Level" => {
+                let val = parse::parse_level(val);
+                self.level.join(Some(val));
+            }
+            "Shortname" => {
+                let val = val.clone();
+                self.shortname.join(Some(val))
+            }
+            "Status" => {
+                let val = val.clone();
+                self.raw_status.join(Some(val));
+            }
+            "Title" => {
+                let val = val.clone();
+                self.title.join(Some(val));
+            }
+            _ => die!("Unknown metadata key \"{}\".", key; line_num),
+        }
+
+        self.has_metadata = true;
+    }
+
+    pub fn fill_macros(&self, doc: &mut Spec) {
+        let macros = &mut doc.macros;
+
+        if let Some(date) = self.date.as_ref() {
```

Comment:
nit: May be better to use `if let Some(ref date) = self.date {` instead and such. A little less noisy IMO.